src/Haskell/Ide/Engine/Dispatcher.hs

```diff
@@ -43,59 +65,92 @@ mainDispatcher inChan ghcChan ideChan = forever $ do
+            in s { requestQueue = newQueue }
 
-ghcDispatcher :: forall void. DispatcherEnv -> TChan GhcRequest -> IdeGhcM void
-ghcDispatcher env@DispatcherEnv{docVersionTVar} pin = forever $ do
+ghcDispatcher :: forall void m. DispatcherEnv -> ErrorHandler -> CallbackHandler m -> TChan (GhcRequest m) -> IdeGhcM void
+ghcDispatcher env@DispatcherEnv{docVersionTVar} errorHandler callbackHandler pin = forever $ do
   debugm "ghcDispatcher: top of loop"
   (GhcRequest context mver mid callback action) <- liftIO $ atomically $ readTChan pin
-  debugm $ "got request with id: " ++ show mid
+  debugm $ "ghcDispatcher:got request with id: " ++ show mid
 
   let runner = case context of
         Nothing -> runActionWithContext Nothing
         Just uri -> case uriToFilePath uri of
           Just fp -> runActionWithContext (Just fp)
           Nothing -> \act -> do
-            debugm "Got malformed uri, running action with default context"
+            debugm "ghcDispatcher:Got malformed uri, running action with default context"
             runActionWithContext Nothing act
 
   let runWithCallback = do
-        r <- runner action
-        liftIO $ callback r
+        result <- runner action
+        liftIO $ case result of
+          IdeResultOk x -> callbackHandler callback x
+          IdeResultFail err ->
+            case mid of
+              Just lid -> errorHandler lid J.InternalError (show err)
+              Nothing -> debugm $ "ghcDispatcher:Got error for a request: " ++ show err 
 
   let runIfVersionMatch = case mver of
         Nothing -> runWithCallback
         Just (uri, reqver) -> do
           curver <- liftIO $ atomically $ Map.lookup uri <$> readTVar docVersionTVar
           if Just reqver /= curver then
-            debugm "not processing request as it is for old version"
+            debugm "ghcDispatcher:not processing request as it is for old version"
           else do
-            debugm "Processing request as version matches"
+            debugm "ghcDispatcher:Processing request as version matches"
             runWithCallback
 
   case mid of
     Nothing -> runIfVersionMatch
-    Just lid -> do
-      cancelled <- liftIO $ atomically $ isCancelled env lid
-      if cancelled
-      then
-        debugm $ "cancelling request: " ++ show lid
-      else do
-        debugm $ "processing request: " ++ show lid
-        runIfVersionMatch
+    Just lid -> checkCancelled env lid errorHandler $ do
+      liftIO $ completedReq env lid
+      runIfVersionMatch
+
+checkCancelled :: MonadIO m => DispatcherEnv -> J.LspId -> ErrorHandler -> m () -> m ()
+checkCancelled env lid errorHandler callback = do
+  cancelled <- liftIO $ atomically isCancelled
+  if cancelled
+    then liftIO $ do
+      -- remove from cancelled and wip list
+      atomically $ modifyTVar' (cancelReqsTVar env) (S.delete lid)
+      completedReq env lid
+      errorHandler lid J.RequestCancelled ""
+    else callback
```

Comment:
If it is not cancelled, we still need to remove it from wipReqs.

Reply:
Non-cancelled requests get removed from wipReqs outside of `checkCancelled`, since we check for cancellation before executing the request and again after executing it
https://github.com/Bubba/haskell-ide-engine/blob/edc5abe06dde60eee11a23b7008d17288ec4bd4f/src/Haskell/Ide/Engine/Dispatcher.hs#L81-L86